File: backend/app/models/testcase.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text, ForeignKey
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship
from datetime import datetime
from app.core.database import Base


class Module(Base):
    """测试模块（树形结构，用于组织用例）"""
    __tablename__ = "modules"

    id = Column(Integer, primary_key=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    name = Column(String(100), nullable=False)
    # 支持子模块（树形结构）
    parent_id = Column(Integer, ForeignKey("modules.id"), nullable=True)
    description = Column(Text)
    created_at = Column(DateTime, default=datetime.utcnow)

    project = relationship("Project", back_populates="modules")
    children = relationship("Module", back_populates="parent")
    parent = relationship("Module", back_populates="children", remote_side=[id])
    test_cases = relationship("ApiTestCase", back_populates="module")


class ApiTestCase(Base):
    """API 测试用例"""
    __tablename__ = "api_test_cases"

    id = Column(Integer, primary_key=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    module_id = Column(Integer, ForeignKey("modules.id"), nullable=True)
    name = Column(String(200), nullable=False)
    description = Column(Text)
    # GET / POST / PUT / DELETE / PATCH etc.
    method = Column(String(10), nullable=False)
    url = Column(String(500), nullable=False)
    headers = Column(JSONB, default={})
    params = Column(JSONB, default={})
    body = Column(Text)
    # json / form / xml / raw
    body_type = Column(String(20), default="json")
    # 断言规则列表
    assertions = Column(JSONB, default=[])
    pre_script = Column(Text)
    post_script = Column(Text)
    # high / medium / low
    priority = Column(String(10), default="medium")
    # draft / active / deprecated
    status = Column(String(20), default="draft")
    created_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    project = relationship("Project", back_populates="test_cases")
    module = relationship("Module", back_populates="test_cases")
    creator = relationship("User", back_populates="created_test_cases")
    results = relationship("TestResult", back_populates="test_case")

File: frontend/src/services/testcase.ts
import { apiClient } from './api'
import type { Module, ApiTestCase, TestRun, TestResult } from '@/types'

export interface TestCasePayload {
  name: string
  description?: string
  module_id?: number | null
  method: string
  url: string
  headers?: Record<string, string>
  params?: Record<string, string>
  body?: string
  body_type?: string
  assertions?: Array<{ type: string; expected: unknown; path?: string }>
  pre_script?: string
  post_script?: string
  priority?: string
  status?: string
}

export interface TestRunPayload {
  name?: string
  environment_id: number
  test_case_ids: number[]
}

export interface PageResult<T> {
  items: T[]
  total: number
}

export const testcaseApi = {
  // 模块
  listModules: (projectId: number) =>
    apiClient.get<Module[]>(`/api/v1/projects/${projectId}/modules`),
  createModule: (projectId: number, data: { name: string; parent_id?: number }) =>
    apiClient.post<Module>(`/api/v1/projects/${projectId}/modules`, data),
  deleteModule: (moduleId: number) =>
    apiClient.delete(`/api/v1/modules/${moduleId}`),

  // 测试用例
  list: (projectId: number, params?: { module_id?: number; status?: string; skip?: number; limit?: number }) =>
    apiClient.get<PageResult<ApiTestCase>>(`/api/v1/projects/${projectId}/testcases`, { params }),
  get: (caseId: number) =>
    apiClient.get<ApiTestCase>(`/api/v1/testcases/${caseId}`),
  create: (projectId: number, data: TestCasePayload) =>
    apiClient.post<ApiTestCase>(`/api/v1/projects/${projectId}/testcases`, data),
  update: (caseId: number, data: Partial<TestCasePayload>) =>
    apiClient.put<ApiTestCase>(`/api/v1/testcases/${caseId}`, data),
  delete: (caseId: number) =>
    apiClient.delete(`/api/v1/testcases/${caseId}`),

  // 测试执行
  createRun: (projectId: number, data: TestRunPayload) =>
    apiClient.post<TestRun>(`/api/v1/projects/${projectId}/testruns`, data),
  listRuns: (projectId: number, params?: { skip?: number; limit?: number }) =>
    apiClient.get<PageResult<TestRun>>(`/api/v1/projects/${projectId}/testruns`, { params }),
  getRun: (projectId: number, runId: number) =>
    apiClient.get<TestRun>(`/api/v1/projects/${projectId}/testruns/${runId}`),
  getRunResults: (projectId: number, runId: number) =>
    apiClient.get<TestResult[]>(`/api/v1/projects/${projectId}/testruns/${runId}/results`),
  deleteRun: (projectId: number, runId: number) =>
    apiClient.delete(`/api/v1/projects/${projectId}/testruns/${runId}`),
}
